### ISSUE TITLE
版本更新到0.0.12，問題還是沒有解決

### ISSUE THREAD

**Issue by s1041613** [riophae/vue-treeselect#46]
@riophae 您好:
我遇到的第一個問題是，單選的下拉選單會出現unknow的預設文字，然後按旁邊的叉叉，才會變成我placeholder預設的文字，我有看您給別人的回答是更新後就可以了，但我的還是在
![q1](https://user-images.githubusercontent.com/35199238/38017965-2cca8fa2-32a6-11e8-8c97-43083a5c5772.jpg)
![q2](https://user-images.githubusercontent.com/35199238/38018071-790ef54c-32a6-11e8-9f80-99c1847bcafa.jpg)
我遇到的第二個問題是，我用延遲加載後，想要預設default，所以塞值在v-model裡面，用console看，值有進去，可是在網頁上看還是沒有，一訂要點擊下拉選單，才會出現預設的選項，沒辦法直接顯示
以上兩個問題，請您幫忙解答，非常感謝!

**Comment by riophae** [comment]
你好！这个 unknown 字样应该是因为你设置了默认选中的选项，但是这个选项没有加载到树里面去，这个和延迟加载有关系。
方便贴一下你调用 vue-treeselect 部分的代码吗？这样有助于我进一步帮到你，谢谢！

**Comment by s1041613** [comment]
@riophae 您好
以下是我的代碼，謝謝!
![11](https://user-images.githubusercontent.com/35199238/38019258-b1d4a04a-32a9-11e8-9491-16b8e7092179.jpg)
![22](https://user-images.githubusercontent.com/35199238/38019264-b6df013e-32a9-11e8-958a-e59f68d8f984.jpg)
![33](https://user-images.githubusercontent.com/35199238/38019270-bb7a55cc-32a9-11e8-8a60-9af668a57387.jpg)

**Comment by riophae** [comment]
我看了一下，没有看到 `optionFlight()` 被调用？而且它被当成参数传给 vue-treeselect 了。请试一下改成下面这样：

```vue
<template>
  <treeselect 
    placeholder="Select..." 
    :max-height="150"
    v-model="responseFlight"
    :load-root-options="loadRootOptions1"
    style="width: 200px;"
    />
</tempalte>

<script>
  export default {
    methods: {
      loadRootOptions1(callback) {
        if (this.apiFlight != '') {
          const newFlight = this.apiFlight.map(d => ({ id: d[0], label: d[1] }))
          callback(null, newFlight)
          this.responseFlight = newFlight[0].id
        }
      },
    },
  }
</script>
```

**Comment by s1041613** [comment]
@riophae 您好
我換成您給的代碼之後，他完全沒有反應，我擺console在if (this.apiFlight != '')之前也沒顯示，可是我在vue的工具裡面看值的時候發現this.apiFlight裡面有值，可是fun動不了，然後我試者直接呼叫loadRootOptions1(callback)，他會跑出callback is undefine,再請您幫忙了，謝謝!

**Comment by riophae** [comment]
似乎是通过另外的函数发起 AJAX 请求，然后再给 `apiFlight` 赋值的对吗？

**Comment by s1041613** [comment]
對

**Comment by riophae** [comment]
嗯，问题应该就在这里了。请在 `loadRootOptions1` 这个函数中，调用发起 AJAX 请求的那个函数；AJAX 得到响应后，再调用 `callback`。

**Comment by s1041613** [comment]
我直接在loadRootOptions1裡面調用，他沒反應，我直接在发起 AJAX 请求的那个函数呼叫loadRootOptions1，就會有問題，我this.apiflight的資料就會進不來
![11](https://user-images.githubusercontent.com/35199238/38069417-c113c8a2-3348-11e8-93ed-864d3109edf9.jpg)
![22](https://user-images.githubusercontent.com/35199238/38069421-c592cc8e-3348-11e8-96d3-e028fde1d760.jpg)
![33](https://user-images.githubusercontent.com/35199238/38069424-ca69311c-3348-11e8-8f90-ca895ecbf9b2.jpg)

**Comment by riophae** [comment]
嗯，是这样的，`loadRootOptions` 这个函数本身是一个 callback 性质的东西，它是由 vue-treeselect 来调用的，不是由你自己的代码来调用的。只有由 vue-treeselect 调用的时候，才会有那个 `callback` 参数。这个 `callback` 参数才是重点……
vue-treeselect 默认会在 mounted 事件的时候调用这个 `loadRootOptions` 函数参数，只有这个时间窗口才允许你通过 `callback` 参数给 vue-treeselect 传递你的选项参数。错过了这个机会就怎么调用都没效果了……

**Comment by riophae** [comment]
假如你有单独的函数在处理 AJAX 请求这一块逻辑的吧，建议把这个函数里的代码移动到 `loadRootOptions1` 里面来，调整成类似这样：

```js
{
  loadRootOptions1(callback) {
    ajax.get('/api', data => {
      callback(null, data)
    })
  },
}
```

**Comment by s1041613** [comment]
@riophae 不好意思想再請教您一個問題，動態載入資料後，將其中一筆綁定vmodel設為預設，在網頁上的呈現方式需要先點擊下拉選單，才可以跑出預設，有辦法在資料載入後，直接顯示預設嗎?謝謝!

**Comment by riophae** [comment]
你的意思是，设置了 v-model 但是没有显示出来吗？这样设置了之后应该已经显示了出来才对，和是否点击下拉选单没关系，有可能是 bug？

**Comment by s1041613** [comment]
![aa](https://user-images.githubusercontent.com/35199238/38077464-9dabbb9a-336b-11e8-99d8-66dff3329638.jpg)
我綁定v-model的選項會在展開之後變成粗體，可是沒辦法一開始就顯示在上面

**Comment by s1041613** [comment]
好的
<template>
           <treeselect
            placeholder="Select..."
            :max-height="150"
            v-model="responseFlight"
            :load-root-options="loadRootOptions1"
            style="width: 200px;"
            />
</template>
<script>
    loadRootOptions1(callback) {
      console.log('in load')
      this.sendAPIReal()
      if (this.apiFlight != ""){
        const newFlight = this.apiFlight.map(d => ({ id: d[0], label: d[1] }))
        callback(null, newFlight)
        this.responseFlight = newFlight[0]
      }
      else {
        setTimeout(() => callback(new Error('test')), 5000)
      }
    },
</script>

**Comment by riophae** [comment]
我明白了，这个很容易解决，做几个小改动就好了：

1. 首先给 `sendApiReal()` 这个函数添加一个 `callback` 参数，变成 `sendApiReal(callback)` 这样
2. 然后在其中 `this.apiFlight = this.dataFilter.flight` 这一行的下面加上一行：`if (callback) callback()`
3. 然后回到 `loadRootOptions1()` 这个函数来，做一些改动，变成这个样子：

```vue
<script>
  export default {
    loadRootOptions1(callback) {
      console.log('in load') 
      this.sendAPIReal(() => { // <- 注意这里
        if (this.apiFlight != "") {
          const newFlight = this.apiFlight.map(d => ({
            id: d[0],
            label: d[1]
          }))
          callback(null, newFlight) 
          this.responseFlight = newFlight[0]
        } else {
          setTimeout(() => callback(new Error('test')), 5000)
        }
      }) 
    },
  }
</script>
```
这样应该就正常了！

**Comment by s1041613** [comment]
我在sendApiReal()裡面按照您說的改了，可是資料就跑不進來了，然後console裡面沒有報錯誤訊息
![4](https://user-images.githubusercontent.com/35199238/38079788-550e1796-3372-11e8-9b0e-b965e384329c.jpg)
![5](https://user-images.githubusercontent.com/35199238/38079795-58fb06ac-3372-11e8-81b0-7f49fdc8c01c.jpg)

**Comment by riophae** [comment]
试试在 `callback(null, newFlight)` 后面加上一句 `console.log(newFlight)`，然后看看 console 里面有没有把数据打印出来呢？正常情况下应该是有打印出来的。如果没有的话，说明 AJAX 请求没发送出去，可能是你在 `sendAPIReal()` 里面最开始的 `if` 语句没能满足条件，所以 `axios.post()` 这里完全没执行……以上只是我的猜测，请先观察一下打印结果看看

**Comment by s1041613** [comment]
sendAPIReal()裡在if之前的console也沒有打印出來，所以他是應該進不去...

**Comment by riophae** [comment]
我觉得是这样的，可能你需要把那个 `if` 条件调整一下

**Comment by s1041613** [comment]
不好意思我想在請問一下，在第三則訊息裡，我貼的代碼，是把sendAPIReal()綁click在button上面的，if通過之後會發送ajax請求，把值給this.apiflight，透過optionFlight這個computed去轉換陣列得到this.newflight，再給loadRootOptions1裡的callback當參數用，這些都是畫面載入後mounted已經結束的動作，可是loadRootOptions1卻可以被調用，下拉裡面的值都有跑出來，請問為甚麼這樣跑得出來..?麻煩您了!!

**Comment by riophae** [comment]
嗯，因为点按钮之后发送了一次 AJAX 请求，除此之外 vue-treeselect 也通过 loadRootOptions1 主动调用过 sendAPIReal() 发送过 AJAX 请求

**Comment by s1041613** [comment]
那請問在loadRootOptions1裡面，把值丟給v-model，想做預設，可是要通過點擊選下拉選單，才可以呈現在網頁上，想請問這部分您是怎麼寫的，因為上面您有說您在那邊測試是正常的~麻煩了，謝謝!!

**Comment by riophae** [comment]
没关系！不过我还有个问题要问，设了 v-model 之后，选项数据没加载下来时，输入框里应该是有文字显示的才对，应该会看到“unknown”字样，加载完成之后“unknown”字样自动消失。现在是点击之前什么文字都没有是吗？连那个“×”都没有吗？

**Comment by riophae** [comment]
没关系！是这样子的，vue-treeselect 有一个 `loadRootOptions` 这个参数对不对？这个参数是 vue-treeselect 会在 `mounted` 事件的时候自动调用的，也就是说就算你自己的代码完全没调用过这个函数，vue-treeselect 也会去调用。然后在这次调用中，会进而调用 `sendAPIReal()` 然后发送 AJAX 请求。
然后回到后面的问题，我觉得这里有点奇怪，照理来说不需要通过点击就应该有选项显示出来了才对。

**Comment by s1041613** [comment]
我後來發現“unknown”字样是因為我一開始在data把綁定的設成responseFlight:""，我把它改成responseFlight:null之後，unknown的字就會消失
現在是點擊之前，出現unknown的字，就算v-model塞值，數據跑完以後，unknown也沒有消失，需要點擊了以後選值才會消失，然後呈現我預設的選項
謝謝!!

**Comment by riophae** [comment]
你应该没有用 vue-treeselect 的 `autoLoadRootOptions` 参数对吧？如果把这个参数值设为 `false`，就会导致等到点开下拉菜单的时候才开始加载选项数据的情况。如果不用这个参数，默认应该是 `mounted` 事件时开始加载。

如果没有用这个参数，但是等到之后点击下拉菜单才看到数据被加载，那么说明第一次 `mounted` 事件时因为某种原因没有成功加载进来。猜测可能是 AJAX 错误。

**Comment by s1041613** [comment]
剛剛發現如果是我第三則訊息的寫法，需要點擊下拉選單才會出發loadRootOptions的樣子，因為我點了之後，console裡面打印出綁定v-model的值
![image](https://user-images.githubusercontent.com/35199238/38129467-bd77ae72-3431-11e8-822f-f6cb3a7870c6.png)

**Comment by s1041613** [comment]
恩恩我沒有寫那行，我第一次mounted不會有結果，因為我的if條件需要等使用者填完有值之後，才會成立，之後才會送出ajax請求，麻煩了

**Comment by riophae** [comment]
看这样行不行，是不是你下面那个下拉选择框没有用到树状选择的功能？如果没有用到的话，建议你换用 [vue-multiselect](https://vue-multiselect.js.org) 这个组件。这个应该说是 vue-treeselect 的问题，暂时来说它对于动态加载选项支持得不够好，所以你写起来就各种预期外的状况，我已经打算去改进这一块了，但是需要花时间。你着急的话，建议先换用到 vue-multiselect，估计用起来会顺手不少。不好意思哈。

**Comment by riophae** [comment]
因为我考虑到，你下面那个下拉框的选项是可能需要多次变动的（如果上面输入框的值发生改动的话），但是 vue-treeselect 暂时不支持你反复更换选项数据，只支持局部数据的延迟加载。这个是实现思路上的限制导致的。

**Comment by s1041613** [comment]
@riophae 好的，非常謝謝你!
